statistics/cmsketch.go

```diff
@@ -44,6 +53,61 @@ func NewCMSketch(d, w int32) *CMSketch {
 	return &CMSketch{depth: d, width: w, table: tbl}
 }
 
+// NewCMSketchWithTopN returns a new CM sketch with TopN elements.
+func NewCMSketchWithTopN(d, w int32, data [][]byte, n int32) *CMSketch {
+	tbl := make([][]uint32, d)
+	for i := range tbl {
+		tbl[i] = make([]uint32, w)
+	}
+	c := &CMSketch{depth: d, width: w, table: tbl}
+	c.BuildTopN(data, n)
+	return c
+}
+
+// BuildTopN builds table of top N elements.
+// elements in data should not be modified after this call.
+func (c *CMSketch) BuildTopN(data [][]byte, n int32) {
+	c.topn = make([]cmscount, n)
+	for k := range data {
+		found, pos := false, -1
+		for i := range c.topn {
+			if bytes.Equal(c.topn[i].data, data[k]) {
+				found = true
+				pos = i
+				break
+			} else if c.topn[i].count == 0 {
+				pos = i
+			}
+		}
+		if found {
+			c.topn[pos].count++
+		} else if !found && pos != -1 {
+			c.topn[pos] = cmscount{data[k], 1}
+		} else {
+			for i := range c.topn {
+				c.topn[i].count--
+			}
+		}
+	}
+	for i := range c.topn {
```

Comment:
If there is no significant higher occurrence rate, for example, at least 5%, there is no need to build top n for the item.